plutus-core/plutus-core/src/PlutusCore/Parsable.hs

```diff
@@ -21,25 +23,97 @@ import           Data.ByteString (ByteString)
 import qualified Data.ByteString as BS (pack)
 import           Data.Char       (ord)
 import qualified Data.Text       as T
-import           Text.Read       (readMaybe)
+import           Text.Read
+
+{- Note [Parsing horribly broken]
+As the title suggests, at the moment parsing is horribly broken. 'parse' expects a closed chunk of
+text, but in order to provide one we need to determine in the main parsing pipeline (which can be
+Happy-based or megaparsec-based) where that closed chunk ends (determining where it starts is easy).
+So we need either of these two:
+
+1. perform lexical analysis, cut the right piece and feed it to 'parse'
+2. make 'parse' take as much as it needs and return the rest to the main parsing machinery
+
+The latter option is quite non-trivial, because we have that Happy parser and it's hard to poke a
+hole in it to get some custom parsing of constants consuming an arbitrary amount of symbols
+(we don't even have symbols in Happy -- it's tokens there). So if we wanted to do the latter,
+it would probably be the easiest option to just remove the Happy parser and replace it with a
+megaparsec-based one (should be the simplest thing in the world, given that PIR is a superset of
+PLC and already has a megaparsec-based parser).
+
+There are arguments in favor of the former option (https://github.com/input-output-hk/plutus/pull/2458#discussion_r522091227):
+
+> If you look in Lexer.x there's extensive commentary about how it parses constants. We could steal
+the relevant regular expressions from there or reuse the entire lexer. I'm not convinced that
+letting types do their own lexical analysis is a good idea: doing these things correctly can be
+quite tricky and if you got it wrong it might mess up the main parser in some subtle way that
+doesn't show up until you encounter an unexpected situation. The way that it's done at the moment
+provides reasonably general syntax for constants and I'm fairly confident it's correct. I guess I
+feel that this is low-level stuff that should be done once and then shut away in a box where it
+won't cause any trouble.
+
+however it's not that simple. Consider a list of lists: let's say the lexer determines what
+constitutes a closed chunk of text representing such a value. And we feed that chunk to 'parse'.
+But now we need to do lexical analysis again, but this time in 'parse' to determine whether a @,@
+is a part of formatting or is inside an element (being, say, a string or a list of strings) of the
+list. I.e. our constant parsers can be recursive and it's a pain to deal with when you do all the
+lexical analysis upfront.
+
+So at the moment we don't do anything correctly. Neither PLC nor PIR can handle lists of lists
+and none of them can handle tuples at all ('Read' provides us with overloaded parsing for lists
```

Comment:
I'm going to pretend that we don't have any problems with the parser within the scope of this PR.

> It would be a lot easier to use some different syntax like `<(con integer 5), (con bool True)>`

(note that it would be `(con (pair integer bool) <5, True>)` then. We can't put arbitrary ASTs inside polymorphic built-in types and `con` creates an "arbitrary AST"). Banning parentheses and square brackets inside the syntax of constants would certainly simplify parsing and I also think that might be a good idea that would probably allow us note to poke a hole the parser and instead parse constants in chunks like we do right now.